tools/depcheck/pkg/cmd/trace/filter.go

```diff
@@ -0,0 +1,143 @@
+package trace
+
+import (
+	"strings"
+
+	"github.com/gonum/graph"
+
+	depgraph "github.com/openshift/origin/tools/depcheck/pkg/graph"
+)
+
+// FilterVendorPackages receives a graph and a set of vendorRootNames contained within the graph.
+// Returns a new graph with the sub-tree for each vendorRoot collapsed into just the vendorRoot.
+// Relationships among vendorRoots are kept: if a transitive closure exists between two
+// vendor sub-trees, an edge is created between their roots in the newly collapsed graph.
+func FilterVendorPackages(g *depgraph.MutableDirectedGraph, vendorRootNames []string) (*depgraph.MutableDirectedGraph, error) {
+	pkgs := &PackageList{
+		Packages: []Package{},
+	}
+
+	vRoots := []*depgraph.Node{}
+
+	for _, n := range g.Nodes() {
+		node, ok := n.(*depgraph.Node)
+		if !ok {
+			continue
+		}
+
+		if !isVendorRoot(node, vendorRootNames) {
+			continue
+		}
+
+		vRoots = append(vRoots, node)
+	}
+
+	for _, r := range vRoots {
+		pkg := Package{
+			ImportPath: r.UniqueName,
+			Imports:    []string{},
+		}
+
+		seenClosures := map[string]bool{}
+		flatDepSet := flatten(g, r)
+
+		for _, n := range flatDepSet {
+			node, ok := n.(*depgraph.Node)
+			if !ok {
+				continue
+			}
+
+			for _, vRoot := range vRoots {
```

Comment:
Not that I won't accept n^3, but it's a little hard to read and it's hard for me to figure out if you've gotten it right.  This feels like it should be n^2 problem on a straight traversal.